### PR TITLE
MM-8607 Add ability to turn off non-critical services when under load

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This pull request adds new functionality to temporarily disable non-critical services when the server is under load.

When a server is experiencing significant load, impacting the user experience, a system admin can use this tool via the system console to temporarily disable non-critical services (search, status notification broadcasts, typing notification broadcasts).  Eventually this "server busy" feature can also be triggered automatically by load detection code.

This pull request provides the backend functionality. A separate PR is needed to add GUI to the system console. 

Specifically, the services deemed non-essential:
- team search
- channel search
- user search
- post search
- status notification broadcasts
- typing notification broadcasts

Three end-points were added, plus client4 APIs, and unit tests.  End-point documentation submitted in separate PR.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-8607

### CODE REVIEW

**Comment by wiggin77** [comment]
Endpoint documentation PR: https://github.com/mattermost/mattermost-api-reference/pull/489

File: api4/system_test.go [hunk 1]
```diff
@@ -507,3 +508,100 @@ func TestRedirectLocation(t *testing.T) {
...
+func TestSetServerBusyInvalidParam(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	t.Run("as system admin, invalid param", func(t *testing.T) {
+		params := []int{-1, 0}
+		for _, p := range params {
+			ok, resp := th.SystemAdminClient.SetServerBusy(p)
+			CheckBadRequestStatus(t, resp)
+			require.Falsef(t, ok, "")
```
Comment (jespino) [inline api4/system_test.go]: I think this line is not needed

File: app/busy_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	require.False(t, busy.IsBusy())
+
+	busy.Set(time.Second * 30)
+	require.True(t, busy.IsBusy())
+
+	expire := busy.Expires()
+	require.Greater(t, expire.Unix(), time.Now().Add(time.Second*10).Unix())
+
+	busy.Clear()
+	require.False(t, busy.IsBusy())
```
Comment (jespino) [inline app/busy_test.go]: Maybe add a case where you set the busy value when is already set (I know is under control, but would be another test case to add).

File: app/busy_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+	isNotBusy := func() bool {
+		return !busy.IsBusy()
+	}
+
+	busy.Set(time.Millisecond * 100)
+	require.True(t, busy.IsBusy())
+	require.Eventually(t, isNotBusy, time.Second*5, time.Millisecond*20)
+
+}
```
Comment (jespino) [inline app/busy_test.go]: Maybe we can add tests about the Expires function, when try to get the expire without setting it, or after setting it.

File: api4/system_test.go (hunk 1 above)
Comment (wiggin77) [reply api4/system_test.go]: Agreed. Removed redundant check.

File: app/busy_test.go (hunk 2 above)
Comment (wiggin77) [reply app/busy_test.go]: Added additional test cases.

File: app/busy_test.go (hunk 3 above)
Comment (wiggin77) [reply app/busy_test.go]: Added additional test cases for getting expiry, including before it was set and after a clear.

**Comment by wiggin77** [comment]
> Looks great, one improvement suggestion:
> You've added some repeating blocks like
> 
> ```
> 	if c.App.Srv.Busy.IsBusy() {
> 		// this is considered a non-critical service and will be disabled when server busy.
> 		c.SetServerBusyError()
> 		return
> 	}
> ```
> 
> Wouldn't it be better if that code was extracted in a 'decorator' function similar to
> `api.ApiSessionRequired`?
> 
> Then all you'll have to do is to decorate the route registration with 'busy check'
> 
> ```
> 	api.BaseRoutes.Team.Handle("/posts/search", api.ApiSessionRequired(api.DisableWhenBusy(searchPosts))).Methods("POST")
> ```

@reflog, thanks for your feedback.  That does look cleaner.  However it does not appear those methods are currently nestable.  I will need to create something like 
```
api.BaseRoutes.Team.Handle("/posts/search", api.ApiSessionRequiredDisableWhenBusy(searchPosts)).Methods("POST")
```
I suspect we'll want to make those nestable if more combinations are needed.

**Comment by reflog** [comment]
@wiggin77 - you are right, they are not composable at the moment. Are you ok with implementing `ApiSessionRequiredDisableWhenBusy` ?

**Comment by wiggin77** [comment]
> @wiggin77 - you are right, they are not composable at the moment. Are you ok with implementing `ApiSessionRequiredDisableWhenBusy` ?

Yes, done via https://github.com/wiggin77/mattermost-server/commit/4ecdd08fcb5e208e90af26eacbcb06c59b24ad7f

File: api4/system.go [hunk 4]
```diff
@@ -444,3 +449,49 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
...
 }
+
+func setServerBusy(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	// number of seconds to keep server marked busy
+	secs := r.URL.Query().Get("secs")
```
Comment (jwilander) [inline api4/system.go]: I think going with `seconds` over `secs` will be clearer in the API

File: api4/system.go [hunk 5]
```diff
@@ -444,3 +449,49 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	if secs == "" {
+		secs = "3600" // default 1 hour
```
Comment (jwilander) [inline api4/system.go]: Might be better as constant such as `DEFAULT_SERVER_BUSY_SECONDS`

File: api4/system.go [hunk 6]
```diff
@@ -444,3 +449,49 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
...
+func getServerBusyExpires(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	m := make(map[string]string)
+	m["busy"] = fmt.Sprintf("%t", c.App.Srv.Busy.IsBusy())
+	m["expires"] = fmt.Sprintf("%d", c.App.Srv.Busy.Expires().Unix())
+	m["expires_ts"] = c.App.Srv.Busy.Expires().UTC().Format("Mon Jan 2 15:04:05 -0700 MST 2006")
```
Comment (jwilander) [inline api4/system.go]: What about making a struct for this? That way the client (in model/client4.go) can decode into that and it's clear to API users what the response structure is

File: api4/system.go [hunk 7]
```diff
@@ -40,6 +41,10 @@ func (api *API) InitSystem() {
 	api.BaseRoutes.ApiRoot.Handle("/redirect_location", api.ApiSessionRequiredTrustRequester(getRedirectLocation)).Methods("GET")
 
 	api.BaseRoutes.ApiRoot.Handle("/notifications/ack", api.ApiSessionRequired(pushNotificationAck)).Methods("POST")
+
+	api.BaseRoutes.ApiRoot.Handle("/svrbusy", api.ApiSessionRequired(setServerBusy)).Methods("POST")
+	api.BaseRoutes.ApiRoot.Handle("/svrbusy", api.ApiSessionRequired(getServerBusyExpires)).Methods("GET")
+	api.BaseRoutes.ApiRoot.Handle("/svrbusy/clear", api.ApiSessionRequired(clearServerBusy)).Methods("POST")
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply api4/system.go]: @jwilander, thanks for your feedback.  I have renamed the endpoints to `server_busy` and will update docs.  https://github.com/mattermost/mattermost-server/pull/13212/commits/d3706fb9ecb362c9d539d373c0f5b2fa147e2442

File: api4/system.go (hunk 5 above)
Comment (wiggin77) [reply api4/system.go]: Done via https://github.com/mattermost/mattermost-server/pull/13212/commits/d3706fb9ecb362c9d539d373c0f5b2fa147e2442

File: api4/system.go (hunk 6 above)
Comment (wiggin77) [reply api4/system.go]: Good idea.  Done via https://github.com/mattermost/mattermost-server/pull/13212/commits/d3706fb9ecb362c9d539d373c0f5b2fa147e2442

File: api4/system.go (hunk 4 above)
Comment (wiggin77) [reply api4/system.go]: Done via https://github.com/mattermost/mattermost-server/pull/13212/commits/d3706fb9ecb362c9d539d373c0f5b2fa147e2442

File: api4/post.go [hunk 8]
```diff
@@ -25,7 +25,7 @@ func (api *API) InitPost() {
 
 	api.BaseRoutes.ChannelForUser.Handle("/posts/unread", api.ApiSessionRequired(getPostsForChannelAroundLastUnread)).Methods("GET")
 
-	api.BaseRoutes.Team.Handle("/posts/search", api.ApiSessionRequired(searchPosts)).Methods("POST")
+	api.BaseRoutes.Team.Handle("/posts/search", api.ApiSessionRequiredDisableWhenBusy(searchPosts)).Methods("POST")
```
Comment (saturninoabril) [inline api4/post.go]: Could you please add "post search" in PR description as pat of none-essential or non-critical services?

File: api4/system.go [hunk 9]
```diff
@@ -444,3 +452,51 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
...
 }
+
+func setServerBusy(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	// number of seconds to keep server marked busy
+	secs := r.URL.Query().Get("seconds")
```
Comment (saturninoabril) [inline api4/system.go]: Should we consider max limit value of "seconds" to prevent setting too high when API is accessed other than webapp?

**Comment by saturninoabril** [comment]
@wiggin77 I missed to mention that this failed due to i18n check - translation string was added but should be in the correct order.  Please take a look, thanks

File: api4/post.go (hunk 8 above)
Comment (wiggin77) [reply api4/post.go]: Good catch.  Added.

File: api4/system.go (hunk 9 above)
Comment (wiggin77) [reply api4/system.go]: I added a maximum via const. Currently set to one day, but easily changed. Via https://github.com/mattermost/mattermost-server/pull/13212/commits/946d716ce21c3cdca1ddf7ab328ea7eb59819519

I also fixed the order for i18n.

**Pull request merged**